### DISCUSSION TITLE
How to manage options for scopes/features

### DISCUSSION THREAD

**Discussion by icing** [curl/curl#14755] [Ideas]
As mentioned in #14578, there is a challenge to apply `CURLOPT_*`to overall operations of libcurl. The example in the issue is about `DoH` requests that inherit all options from a transfer without that being intended. The reverse is also true, the user is unable to set options for `DoH` requests without affecting the transfer as well.

### libcurl API

It seems unfeasible to add special `CURLOPT_*` variations for all of these. The number of options would just explode. Instead, another idea has been floated: use an easy handle to set options for a specific scope/aspect. Like via:

```
doh = curl_easy_init();
curl_easy_setopt(doh, CURLOPT_URL, CURLOPT_TIMEOUT, 10);
rc = curl_share_setopt(share, CURLSHOPT_DOH, doh); 
curl_easy_cleanup(doh);
```

or, in a multi handle application:

```
doh = curl_easy_init();
curl_easy_setopt(doh, CURLOPT_URL, CURLOPT_TIMEOUT, 10);
rc = curl_multi_setopt(multi, CURLMOPT_DOH, doh); 
curl_easy_cleanup(doh);
```

Internally, the share/multi makes a copy of the passed easy handle and keeps it around. When a DoH transfer is started, it looks for an "easy doh template" and uses that - no longer inheriting options from the transfer starting the DoH request.

Besides DoH, there are other parts where the same pattern can be applied. For example, the internal janitor handle doing connection shutdowns/maintenance.

### curl command line

It would seem desirable to use the same solution on the command line as well. Something like

```
curl --doh-start <options for doh handles> --doh-end
```
or something like:
```
curl --doh "<options for doh handles>"
```

The first seems clean, but somewhat unelegant. The second is like a command line in the command line, but suffers from quoting and var substitution confusion. Maybe there is a good third way of addressing this?

**Comment by vszakats** [comment]
Perhaps curl could automatically recognize `--doh-` as an _option prefix_ for all shareable options, and internally handle them all as 'options for doh handles'? This specific prefix matches what's used now, so existing options would continue to work, while automagically being extended to every other "shareable" option?

In the documentation it would be enough to keep a single copy/instance for such options, while saying which prefixes it supports. E.g. `--doh-`, `--proxy-`, etc.

This method would also leave options position-agnostic.

**Reply by bziemons** [reply]
I would prefer this suggestion, because I cannot imagine wide-spread tab completion to work well with the original suggestions. I use tab completion especially for options I rarely use, so I imagine this to be the case here.

**Comment by bagder** [comment]
I'm in favor. I can imagine adding a `CURLOPT_DOH` option as the first easiest start, to just set it for a specific transfer. Having it shared in the shared object or multi handle feels like a next step.

Maybe we should add a suffix/prefix or something so that we then can use the same suffix when we support this kind of handle for other things? (like proxy or shutdown handle) Maybe `_HANDLE`, `_ANCHOR`, `_COPY` or something else?

For each option that is supported for this DoH handle, we mark it somehow in the libcurl/opts/[option].md markdown header so that we can generate a list easy for which options that are supported. Not supported options should probably just be ignored when set (ie not return an error).

We should probably start off with trying this on the libcurl code only, waiting with adapting the curl tool to it. Because the tool has a completely different set of challenges for both code, UI and documentation.

**Comment by icing** [comment]
Closing due to lack of interest.
